src/codegen/sys/tests.rs

```diff
@@ -287,98 +342,86 @@ fn generate_abi_rs(
-    writeln!(w, "use {}::*;\n", crate_name)?;
-    writeln!(w, "static PACKAGES: &[&str] = &[\"{}\"];", package_name)?;
+    writeln!(w, "use {}::*;", crate_name)?;
     writeln!(
         w,
         "{}",
         r####"
-#[derive(Clone, Debug)]
-struct Compiler {
-    pub args: Vec<String>,
-}
-
-impl Compiler {
-    pub fn new() -> Result<Compiler, Box<dyn Error>> {
-        let mut args = get_var("CC", "cc")?;
-        args.push("-Wno-deprecated-declarations".to_owned());
-        // For _Generic
-        args.push("-std=c11".to_owned());
-        // For %z support in printf when using MinGW.
-        args.push("-D__USE_MINGW_ANSI_STDIO".to_owned());
-        args.extend(get_var("CFLAGS", "")?);
-        args.extend(get_var("CPPFLAGS", "")?);
-        args.extend(pkg_config_cflags(PACKAGES)?);
-        Ok(Compiler { args })
+mod c_abi {
+    #[derive(Copy, Clone, Debug, Eq, PartialEq)]
+    #[repr(C)]
+    pub struct CConstant {
+        pub name: *const libc::c_char,
+        pub value: *const libc::c_char,
     }
 
-    pub fn compile(&self, src: &Path, out: &Path) -> Result<(), Box<dyn Error>> {
-        let mut cmd = self.to_command();
-        cmd.arg(src);
-        cmd.arg("-o");
-        cmd.arg(out);
-        let status = cmd.spawn()?.wait()?;
-        if !status.success() {
-            return Err(format!("compilation command {:?} failed, {}", &cmd, status).into());
-        }
-        Ok(())
+    #[derive(Copy, Clone, Debug, Eq, PartialEq)]
+    #[repr(C)]
+    pub struct CLayout {
+        pub name: *const libc::c_char,
+        pub size: usize,
+        pub alignment: usize,
     }
 
-    fn to_command(&self) -> Command {
-        let mut cmd = Command::new(&self.args[0]);
-        cmd.args(&self.args[1..]);
-        cmd
+    extern "C" {
+        pub fn c_constants(n: *mut usize) -> *mut CConstant;
+        pub fn c_constants_free(c: *mut CConstant, n: usize);
+        pub fn c_layouts(n: *mut usize) -> *const CLayout;
     }
 }
 
-fn get_var(name: &str, default: &str) -> Result<Vec<String>, Box<dyn Error>> {
-    match env::var(name) {
-        Ok(value) => Ok(shell_words::split(&value)?),
-        Err(env::VarError::NotPresent) => Ok(shell_words::split(default)?),
-        Err(err) => Err(format!("{} {}", name, err).into()),
-    }
-}
+fn c_constants() -> Vec<(String, String)> {
+    let mut res: Vec<(String, String)> = Vec::new();
 
-fn pkg_config_cflags(packages: &[&str]) -> Result<Vec<String>, Box<dyn Error>> {
-    if packages.is_empty() {
-        return Ok(Vec::new());
-    }
-    let pkg_config = env::var_os("PKG_CONFIG")
-        .unwrap_or_else(|| OsString::from("pkg-config"));
-    let mut cmd = Command::new(pkg_config);
-    cmd.arg("--cflags");
-    cmd.args(packages);
-    let out = cmd.output()?;
-    if !out.status.success() {
-        return Err(format!("command {:?} returned {}",
-                           &cmd, out.status).into());
+    unsafe {
+        let mut n = 0;
```

Comment:
Or `mem::MaybeUninit` if you feel like it

Reply:
dunno... just init to 0 seem easier/cleaner, but no strong opinion